File: tools/claims_lint.py
#!/usr/bin/env python3
"""
KPI: Every claim in spec/claims.yaml must have metric, baseline, target, gate, owner, benchmark_id, and kill_condition.
"""

import sys
from pathlib import Path

def parse_claims_simple(text: str) -> list:
    claims = []
    current = None
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        if line.startswith("- id:"):
            if current:
                claims.append(current)
            current = {"id": line.split(":", 1)[1].strip().strip('"')}
        elif current and ":" in line:
            k, v = line.split(":", 1)
            current[k.strip()] = v.strip().strip('"')
    if current:
        claims.append(current)
    return claims

def main():
    spec_file = Path(__file__).resolve().parent.parent / "spec" / "claims.yaml"
    print(f"[CLAIMS LINT PY] Validating claim ledger at: {spec_file}")
    
    if not spec_file.exists():
        print(f"[FAIL] Missing {spec_file}")
        sys.exit(1)
        
    text = spec_file.read_text(encoding="utf-8")
    claims = parse_claims_simple(text)
        
    required_keys = ["id", "description", "metric", "baseline", "target", "gate", "owner", "benchmark_id", "kill_condition"]
    errors = []
    
    for claim in claims:
        cid = claim.get("id", "UNKNOWN")
        for k in required_keys:
            if k not in claim or claim[k] is None:
                errors.append(f"Claim '{cid}' is missing required field '{k}'")
        if "POST-FRONTIER" in cid and not claim.get("benchmark_id"):
            errors.append(f"Post-frontier claim '{cid}' must resolve to a valid benchmark_id!")
                
    if errors:
        print(f"[FAIL] {len(errors)} errors found in claim ledger:")
        for err in errors:
            print(f"  - {err}")
        sys.exit(1)
    else:
        print(f"[PASS] All {len(claims)} claims in ledger are valid and falsable.")
        sys.exit(0)

if __name__ == "__main__":
    main()

File: tools/claims_lint.rs
// INVARIANT: Every quantitative claim in spec/claims.yaml must be falsable and resolve to a benchmark ID.
// KPI: 100% of claims have metric, baseline, target, gate, owner, kill_condition, and benchmark_id.

use std::fs;
use std::path::Path;
use std::process::exit;

#[derive(Debug, Default)]
struct Claim {
    id: Option<String>,
    description: Option<String>,
    metric: Option<String>,
    baseline: Option<String>,
    target: Option<String>,
    gate: Option<String>,
    owner: Option<String>,
    benchmark_id: Option<String>,
    kill_condition: Option<String>,
}

fn parse_claims_yaml(content: &str) -> Vec<Claim> {
    let mut claims = Vec::new();
    let mut current: Option<Claim> = None;

    for line in content.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }

        if trimmed.starts_with("- id:") {
            if let Some(c) = current.take() {
                claims.push(c);
            }
            let mut c = Claim::default();
            c.id = Some(trimmed.trim_start_matches("- id:").trim().trim_matches('"').to_string());
            current = Some(c);
        } else if let Some(ref mut c) = current {
            if let Some((key, val)) = trimmed.split_once(':') {
                let key = key.trim();
                let val = val.trim().trim_matches('"').to_string();
                match key {
                    "description" => c.description = Some(val),
                    "metric" => c.metric = Some(val),
                    "baseline" => c.baseline = Some(val),
                    "target" => c.target = Some(val),
                    "gate" => c.gate = Some(val),
                    "owner" => c.owner = Some(val),
                    "benchmark_id" => c.benchmark_id = Some(val),
                    "kill_condition" => c.kill_condition = Some(val),
                    _ => {}
                }
            }
        }
    }

    if let Some(c) = current {
        claims.push(c);
    }

    claims
}

fn main() {
    let spec_path = Path::new("spec/claims.yaml");
    println!("[CLAIMS LINT RS] Auditing claim ledger at: {:?}", spec_path);

    if !spec_path.exists() {
        eprintln!("[FAIL] Missing spec/claims.yaml file!");
        exit(1);
    }

    let content = fs::read_to_string(spec_path).expect("Failed to read spec/claims.yaml");
    let claims = parse_claims_yaml(&content);

    let mut errors = Vec::new();

    for claim in &claims {
        let cid = claim.id.as_deref().unwrap_or("UNKNOWN");

        // KPI: every quantitative claim must resolve to a benchmark id.
        if claim.metric.is_none() {
            errors.push(format!("Claim '{}' is missing field 'metric'", cid));
        }
        if claim.baseline.is_none() {
            errors.push(format!("Claim '{}' is missing field 'baseline'", cid));
        }
        if claim.target.is_none() {
            errors.push(format!("Claim '{}' is missing field 'target'", cid));
        }
        if claim.gate.is_none() {
            errors.push(format!("Claim '{}' is missing field 'gate'", cid));
        }
        if claim.owner.is_none() {
            errors.push(format!("Claim '{}' is missing field 'owner'", cid));
        }
        if claim.benchmark_id.is_none() {
            errors.push(format!("Claim '{}' is missing field 'benchmark_id'", cid));
        }
        if claim.kill_condition.is_none() {
            errors.push(format!("Claim '{}' is missing field 'kill_condition'", cid));
        }

        if cid.contains("POST-FRONTIER") && claim.benchmark_id.is_none() {
            errors.push(format!("Post-frontier claim '{}' must resolve to a valid benchmark_id!", cid));
        }
    }

    if !errors.is_empty() {
        eprintln!("[FAIL] Found {} validation errors in claim ledger:", errors.len());
        for err in &errors {
            eprintln!("  - {}", err);
        }
        exit(1);
    }

    println!(
        "[PASS] All {} claims in ledger are valid, falsable, and resolved to benchmark IDs.",
        claims.len()
    );
}
